rollout/trafficrouting.go

```diff
@@ -17,152 +17,177 @@ import (
+		}
 	}
 	if rollout.Spec.Strategy.Canary.TrafficRouting.Ambassador != nil {
 		ac := ambassador.NewDynamicClient(c.dynamicclientset, rollout.GetNamespace())
-		return ambassador.NewReconciler(rollout, ac, c.recorder), nil
+		trafficReconcilers = append(trafficReconcilers, ambassador.NewReconciler(rollout, ac, c.recorder))
 	}
+
+	// ensure that the trafficReconcilers is a healthy list and its not empty
+	if len(trafficReconcilers) > 0 {
+		return trafficReconcilers, nil
+	}
+
 	return nil, nil
 }
 
 func (c *rolloutContext) reconcileTrafficRouting() error {
-	reconciler, err := c.newTrafficRoutingReconciler(c)
+	reconcilers, err := c.newTrafficRoutingReconciler(c)
+	// a return here does ensure that all trafficReconcilers are healthy
+	// and sane in syntax
 	if err != nil {
 		return err
 	}
-	if reconciler == nil {
+	// ensure that trafficReconcilers list is healthy
+	if len(reconcilers) == 0 {
+		c.log.Info("No TrafficRouting Reconcilers found")
 		return nil
 	}
-	c.log.Infof("Reconciling TrafficRouting with type '%s'", reconciler.Type())
+	c.log.Infof("Found %d TrafficRouting Reconcilers", len(reconcilers))
+	// iterate over the list of trafficReconcilers
+	for _, reconciler := range reconcilers {
+		c.log.Infof("Reconciling TrafficRouting with type '%s'", reconciler.Type())
 
-	var canaryHash, stableHash string
-	if c.stableRS != nil {
-		stableHash = c.stableRS.Labels[v1alpha1.DefaultRolloutUniqueLabelKey]
-	}
-	if c.newRS != nil {
-		canaryHash = c.newRS.Labels[v1alpha1.DefaultRolloutUniqueLabelKey]
-	}
-	err = reconciler.UpdateHash(canaryHash, stableHash)
-	if err != nil {
-		return err
-	}
+		var canaryHash, stableHash string
+		if c.stableRS != nil {
+			stableHash = c.stableRS.Labels[v1alpha1.DefaultRolloutUniqueLabelKey]
+		}
+		if c.newRS != nil {
+			canaryHash = c.newRS.Labels[v1alpha1.DefaultRolloutUniqueLabelKey]
+		}
+		err = reconciler.UpdateHash(canaryHash, stableHash)
+		if err != nil {
+			return err
+		}
 
-	currentStep, index := replicasetutil.GetCurrentCanaryStep(c.rollout)
-	desiredWeight := int32(0)
-	weightDestinations := make([]trafficrouting.WeightDestination, 0)
-	if rolloututil.IsFullyPromoted(c.rollout) {
-		// when we are fully promoted. desired canary weight should be 0
-	} else if c.pauseContext.IsAborted() {
-		// when promote aborted. desired canary weight should be 0
-	} else if c.newRS == nil || c.newRS.Status.AvailableReplicas == 0 {
-		// when newRS is not available or replicas num is 0. never weight to canary
-	} else if index != nil {
-		atDesiredReplicaCount := replicasetutil.AtDesiredReplicaCountsForCanary(c.rollout, c.newRS, c.stableRS, c.otherRSs)
-		if !atDesiredReplicaCount {
-			// Use the previous weight since the new RS is not ready for a new weight
-			for i := *index - 1; i >= 0; i-- {
-				step := c.rollout.Spec.Strategy.Canary.Steps[i]
-				if step.SetWeight != nil {
-					desiredWeight = *step.SetWeight
-					break
+		currentStep, index := replicasetutil.GetCurrentCanaryStep(c.rollout)
```

Comment:
even this block isnt changed..fix indentation

Reply:
this is some visual effect of github diff.
open the file in ur IDE, you won't see any indentation issue.
Line 97 adds a loop, where we have to iterate over the list of trafficReconcilers.
indent is right